src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorMapToDocumentEditsEndpoint.cs

```diff
@@ -0,0 +1,142 @@
+
+                if (request.Kind == RazorLanguageKind.CSharp)
+                {
+                    UnwrapCSharpSnippets(mappedEdits);
+                }
+
+                return new RazorMapToDocumentEditsResponse()
+                {
+                    TextEdits = mappedEdits,
+                    HostDocumentVersion = documentContext.Version,
+                };
+            }
+
+            if (request.Kind != RazorLanguageKind.CSharp)
+            {
+                // All other non-C# requests map directly to where they are in the document.
+                return new RazorMapToDocumentEditsResponse()
+                {
+                    TextEdits = request.ProjectedTextEdits,
+                    HostDocumentVersion = documentContext.Version,
+                };
+            }
+
+            var documentMappingService = requestContext.GetRequiredService<RazorDocumentMappingService>();
+            var edits = new List<TextEdit>();
+            for (var i = 0; i < request.ProjectedTextEdits.Length; i++)
+            {
+                var projectedRange = request.ProjectedTextEdits[i].Range;
+                if (!documentMappingService.TryMapFromProjectedDocumentRange(codeDocument, projectedRange, out var originalRange))
+                {
+                    // Can't map range. Discard this edit.
+                    continue;
+                }
+
+                var edit = new TextEdit()
+                {
+                    Range = originalRange,
+                    NewText = request.ProjectedTextEdits[i].NewText
+                };
+
+                edits.Add(edit);
+            }
+
+            return new RazorMapToDocumentEditsResponse()
+            {
+                TextEdits = edits.ToArray(),
+                HostDocumentVersion = documentContext.Version,
+            };
+
+            static void WrapCSharpSnippets(TextEdit[] snippetEdits)
+            {
+                for (var i = 0; i < snippetEdits.Length; i++)
+                {
+                    var snippetEdit = snippetEdits[i];
+
+                    // Formatting doesn't work with syntax errors caused by the cursor marker ($0).
+                    // So, let's avoid the error by wrapping the cursor marker in a comment.
+                    var wrappedText = snippetEdit.NewText.Replace("$0", "/*$0*/");
+                    snippetEdit.NewText = wrappedText;
+                }
+            }
+
+            static void UnwrapCSharpSnippets(TextEdit[] snippetEdits)
+            {
+                for (var i = 0; i < snippetEdits.Length; i++)
+                {
+                    var snippetEdit = snippetEdits[i];
+
+                    // Unwrap the cursor marker.
+                    var unwrappedText = snippetEdit.NewText.Replace("/*$0*/", "$0");
```

Comment:
I know this isn't your code, but its code I haven't seen it before, and its wild that this works. Doesn't this mean formatting edits that occur after the cursor might be off by 4 characters?!